### PR TITLE
feat: add google cloud source repo support (#7534)

### LINKED ISSUE #7534: Add support for Google Cloud Source repositories
# Summary

This enhancement aims at adding support for Google Cloud Source repositories. There are already a couple of issues (see #3379 and #2814), where the overall consensus was that it is currently not possible or experimental at best. Also, there is no documented and easy to use solution to clone Google Cloud Source repositories.

# Motivation

ArgoCD users with repositories hosted on Google Cloud Source are currently unable to use these repositories with ArgoCD. Adding support for such repositories would allow such users to use ArgoCD as well.

# Proposal

I am proposing support for Google Cloud Service Account credentials in JSON format as part of one of the ArgoCD repository or repository credential secrets, which can be used to create access tokens to clone Google Cloud Source repositories.

I already have a locally tested solution for this. I am planning to refactor a couple of bits, create proper documentation, argocd-cli support, etc. and then to create a PR referencing this enhancement proposal.

### PR DESCRIPTION
**Pull request by alexef**
Screenshot 2022-12-08 at 16 50 39

fixes: #7534

a rebase of #7535 (and #11543) with latest changes from master

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by alexef** [comment]
@crenshaw-dev so I was able to test this on my laptop, created both a repo and a repo credential, created the application, everything works as expected (see screenshots)
Screenshot 2022-12-09 at 13 18 35
Screenshot 2022-12-09 at 13 19 32
Screenshot 2022-12-09 at 13 19 51

**Comment by alexef** [comment]
Failing e2e seems unrelated.

File: ui/src/app/settings/components/repos-list/repos-list.tsx [hunk 1]
```diff
@@ -81,7 +90,13 @@ interface NewGitHubAppRepoCredsParams {
 export enum ConnectionMethod {
     SSH = 'via SSH',
     HTTPS = 'via HTTPS',
-    GITHUBAPP = 'via GitHub App'
+    GITHUBAPP = 'via GitHub App',
+    GOOGLECLOUD = 'via Google Cloud'
+}
+
+interface NewGoogleCloudSourceRepoCredsParams {
```
Comment (keithchong) [inline ui/src/app/settings/components/repos-list/repos-list.tsx:97]: All the UI changes look good - very similar to the other connection type and credentials.  The only thing I can nitpick on is just to move this interface above the enum so all the interfaces are 'together'.

Comment (alexef) [reply ui/src/app/settings/components/repos-list/repos-list.tsx]: updated!

**Comment by keithchong** [comment]
Hi @crenshaw-dev , LGTM for the frontend changes.  I presume you will review the other changes?

**Review by crenshaw-dev — APPROVED** [review]
lgtm!

One nitpick if you have time: the images are a bit big, and folks have been concerned about repo size. Can we take 'em down to ~1000px on the longest dimension and run them through something like oxiPng to compress?

**Review by keithchong — APPROVED** [review]

**Comment by crenshaw-dev** [comment]
@alexef can you rebase? It'll include a fix for those nasty e2e test failures.

**Comment by alexef** [comment]
@crenshaw-dev sure thing

**Comment by alexef** [comment]
@crenshaw-dev updated.

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**